I've updated the references in Cora, Locker and Asset

jcfrei opened the pull request:
Cora now has a Set<Locker>, and Locker has a reference to a Cora
Locker has Set<Asset>, and Asset has a reference to a Locker
LockerAsset has been deleted

not yet implemented:
Company with a Set<Cora>, and Cora with a reference to a Company is still outstanding


**Please check if these changes break any existing plans/implementations for our repositories**
Also, I tried to implement a reference inside Cora to a Company but received the following exception:

org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'masterEntityManager' defined in class path resource [com/pidas/cora/config/DatabaseConfiguration.class]: Invocation of init method failed; nested exception is org.hibernate.AnnotationException: Use of @OneToMany or @ManyToMany targeting an unmapped class: com.pidas.cora.model.master.Company.coras[com.pidas.cora.model.tenant.Cora]

I couldn't yet figure out what the problem is - has it to do with the packages?
I believe we should be able to retrieve a List of Coras given a specific company.

msh-wtag commented:
Dear @jcfrei ,
Regarding the error you found during mapping between Cora and Company, 
Company is in master schema and Cora is in tenant schemas, so mapping between Company and Cora is not possible in that way. We urge you to arrange a meeting regarding the structure.

Another point regarding Locker and Asset mapping, we added LockerAsset mapping because it has manyToMany relation and it has extra field quantity between Locker and Asset e.g. one locker can have multiple asset of same type. We need to understand how you want to implement the feature? 

Again, we think you need to sit and discuss about those and clarify our understating about above mentioned structure ASAP.

jcfrei commented:
> Regarding the error you found during mapping between Cora and Company,
Company is in master schema and Cora is in tenant schemas, so mapping between Company and Cora is not possible in that way. We urge you to arrange a meeting regarding the structure.

thanks for explaining. there needs to be the possibility to assign coras to companies. if that's possible with the current structure, then we are fine. ie. there needs to be a getCorasOfCompany(String company) method.

> Another point regarding Locker and Asset mapping, we added LockerAsset mapping because it has manyToMany relation and it has extra field quantity between Locker and Asset e.g. one locker can have multiple asset of same type. We need to understand how you want to implement the feature?

The relationship between Asset and Lockers is ManyToOne from my understanding. Many different Asset instances can be assigned to one Locker. for example a power supply and a laptop. I believe the field Set<Asset> assets should therefore be a List and not a Set - I've corrected my code accordingly.

So just to clarify another more concrete example to show what should be possible:
We have a table (on mssql) for assets like this
id, locker_id, assetType ...
1,  1,  "ethernet cable"
2,  1,  "ethernet cable"
3,  2,  "ethernet cable"

so we have two ethernet cables in locker 1.

Please tell me if this makes sense and seems reasonable to you.

msh-wtag commented:
Dear @jcfrei ,
We had meeting with Philipp, from there we understood that one locker can have only one asset type, which was unknown to us. For this reason I have removed LockerAsset mapping. I am implementing the mapping in different way. 

Asset model will be AssetType and there will be another model named Asset which will be mapped OneToOne with Locker as one locker.

I have committed a version into my branch PIDAS-CORA-SUPPORT-TASK. Please don't merge it yes as it is incomplete. 
Please check if model structure is simpler than previous.

jcfrei commented:
Thanks! 

I've just talked with Philipp to clarify this further:
On the database level we want to stay as flexible as possible. Meaning a locker can hold an (almost) infinite amount of different assets (ie. with different asset types). On the application level we will limit what tasks a user can perform using Rules.  we suggest adding a default rule which allows one locker to hold only assets with one assetType (however this rule can be deactivated). the rule would allow any an asset of any assetType to be added to an **empty** locker. However after on asset has been deposited to that locker, only further Assets of the same assetType can be added.


Regarding assets and lockers I am still confused about their relationship. If I create a two new asset objects for a power supply, they will be unique with regards to their id. Then I can add both of these asset objects to one locker. If we need the quantity of assets within one locker we simply return the size of the list. If we need the number of assets of one particular type, then we iterate through that list and check if it's of the requested type. 
Now regarding your commit:
I would suggest using OneToMany inside of Locker for the Asset List
And ManyToOne for the field Locker inside of Asset

Asset should not required a field lockerId but an assetId which is autogenerated. 
No quantities field.

I would also suggest:
ManyToOne inside of Asset for the assetType, asset is the owning entity, that means one asset has a database entry that references one asset type
OneToMany inside of AssetType for Asset

I don't understand why Asset has a field quantity and why asset is mapped OneToOne to locker.
I would prefer if we could assign infinitely many assets of the same asset type to one locker. For example a locker holds five assets, of which each one is of type "ethernet cable". 

does this make sense? otherwise we can arrange for a call tomorrow afternoon and discuss this.

msh-wtag commented:
Yes I understood your point. So you suggest to add any asset to a locker individually?
Suppose, I want to store 3 mouse(same asset type) to a locker. Then we will store 3 mouse as 3 entry into Asset? And these 3 Asset will be mapped to a Locker right? No quantity field is necessary then.

Thank you for your advice 👍

jcfrei commented:
Yes, exactly. I think if we keep our data structure as close as possible to reality and map one asset object to one actual mouse then we can stay as flexible as possible.
